---
title: "[Next.js Doc] 08. Error Handling"
date: 2024-01-12
description: Next.js Doc - Error Handling
---

#### 참고 문서

[Error Handling – Next.js doc](https://nextjs.org/docs/app/building-your-application/routing/error-handling)

#### On this page

`error.js` 파일 규칙을 사용하면 [중첩된 라우트](https://nextjs.org/docs/app/building-your-application/routing#nested-routes)에서 예상하지 못한 런타임 오류를 적절하게 처리할 수 있다.

- [React Error Boundary](https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary) 에서 라우트 세그먼트와 중첩된 하위 항목을 자동으로 래핑한다.
- 세분성을 조정하기 위해 파일 시스템 계층 구조를 사용하여 특정 세그먼트에 맞는 맞춤형 오류 UI를 만든다.
- 나머지 애플리케이션 기능을 유지하면서 영향을 받은 세그먼트에 대한 오류를 격리한다.
- 전체 페이지를 다시 로드하지 않고 오류 복구를 시도하는 기능을 추가한다.

라우트 세그먼트 안에 `error.js` 파일을 추가하고 React 컴포넌트를 export하여 오류 UI를 만든다:
![image](https://github.com/tamoimi/tami-dev-blog/assets/100749520/b93e69d4-0c73-4bb7-b5ef-8d2203a26119)

`app/dashboard/error.tsx`

```tsx
"use client"; // Error components must be Client Components

import { useEffect } from "react";

export default function Error({ error, reset }: { error: Error & { digest?: string }; reset: () => void }) {
  useEffect(() => {
    // Log the error to an error reporting service
    console.error(error);
  }, [error]);

  return (
    <div>
      <h2>Something went wrong!</h2>
      <button
        onClick={
          // Attempt to recover by trying to re-render the segment
          () => reset()
        }
      >
        Try again
      </button>
    </div>
  );
}
```

---

#### `error.js` 작동 방법

![image](https://github.com/tamoimi/tami-dev-blog/assets/100749520/039ae9ad-4f46-4ba4-ac1a-b34489c9659c)

- `error.js`는 중첩된 자식 세그먼트 또는 `page.js` 컴포넌트를 **래핑**하는 [React Error Boundary](https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary)를 자동으로 생성한다.
- `error.js` 파일에서 내보낸 React 컴포넌트가 **fallback** 컴포넌트로 사용된다.
- error boundary 내에서 오류가 발생하면 오류가 **포함**되고 fallback 컴포넌트가 **렌더링**된다.
- fallback error 컴포넌트가 활성화되면 error boundary **위**의 레이아웃은 상태를 **유지**하고 대화형을 **유지**하며 오류 컴포넌트는 오류를 복구하는 기능을 표시할 수 있다.

---

#### 오류 복구

오류의 원인은 일시적일 수 있다. 이 경우 다시 시도하면 문제가 해결될 수도 있다.

오류 컴포넌트는 `reset()` 함수를 사용해 사용자에게 오류 복구를 시도하라는 메시지를 표시할 수 있다.

이것이 실행될 때 함수는 Error boundary의 내용을 리렌더링하려고 한다. 만약 성공하면 fallback 오류 컴포넌트가 다시 렌더링된 결과로 대체된다.

`app/dashboard/error.tsx`

```tsx
"use client";

export default function Error({ error, reset }: { error: Error & { digest?: string }; reset: () => void }) {
  return (
    <div>
      <h2>Something went wrong!</h2>
      <button onClick={() => reset()}>Try again</button>
    </div>
  );
}
```

---

#### 중첩 라우트

[특수 파일](https://nextjs.org/docs/app/building-your-application/routing#file-conventions)로 생성된 React 컴포넌트는 [특정 중첩 계층 구조](https://nextjs.org/docs/app/building-your-application/routing#component-hierarchy)로 렌더링된다.

예를 들어 `layout.js` 및 `error.js` 파일 모두를 포함하는 두 개의 세그먼트가 있는 중첩 라우트는 아래와 같은 _단순화된_ 컴포넌트 계층 구조로 렌더링된다:

![image](https://github.com/tamoimi/tami-dev-blog/assets/100749520/866429f0-6afb-43eb-809c-c6608c3489fb)
중첩된 컴포넌트 계층 구조는 중첩된 경로에서 `error.js` 파일의 동작에 영향을 미친다:

- 오류는 가장 가까운 부모 error boundary까지 표시된다. 즉 `error.js` 파일이 중첩된 모든 자식 세그먼트에 대한 오류를 처리한다는 뜻이다. 라우트의 중첩 폴더에서 서로 다른 수준의 `error.js` 파일을 배치하면 세분화된 오류 UI를 얻을 수 있다.
- 오류 boundary는 해당 레이아웃의 컴포넌트 **내에** 중첩되어 있기에 `error.js` boundary는 **같은** 세그먼트의 `layout.js` 컴포넌트에서 발생한 오류를 처리하지 **않는다**.

---

#### 레이아웃 오류 처리

`error.js` boundaries는 **같은 세그먼트**의 `layout.js` 또는 `template.js` 컴포넌트에서 발생하는 오류를 캐치하지 **않는**다. 이 [의도적 계층 구조](https://nextjs.org/docs/app/building-your-application/routing/error-handling#nested-routes)는 오류가 발생할 때 형제 라우트(예: 탐색) 간에 공유되는 중요한 UI를 표시하고 작동하도록 유지한다.

특정 레이아웃이나 템플릿 내의 오류를 처리하려면 `error.js` 파일을 레이아웃의 부모 세그먼트에 배치하자.

루트 레이아웃이나 템플릿 내의 오류를 처리하려면 `global-error.js`라는 `error.js`의 변형을 사용하자.

---

#### Root 레이아웃 오류 처리

Root `app/error.js` boundary는 root `app/layout.js` 또는 `app/template.js` 컴포넌트에서 발생한 오류를 캐치하지 **않는**다.

Root 컴포넌트의 오류를 구체적으로 처리하려면 root `app` 디렉터리에 있는 `app/global-error.js`라는 `error.js` 변형을 사용하자.

Root `error.js`와 달리 `global-error.js` 오류 boundary는 **전체** 앱을 감싸고 해당 fallback 컴포넌트는 활성화되면 root 레이아웃을 대체한다. 이 때문에 `global-error.js`는 자체 `<html>` 및 `<body>` **태그를 정의**해야 한다는 점에 유의하는 것이 중요하다.

`global-error.js`는 가장 세분화되지 않은 오류 UI이며 전체 앱에 대한 "catch-all" 오류 처리로 간주될 수 있다. root 컴포넌트는 일반적으로 덜 다이나믹하며 다른 `error.js` 경계는 대부분의 오류를 캐치하므로 자주 트리거될 가능성이 없다.

`global-error.js`가 정의된 경우에도 전 세계적으로 공유되는 UI 및 브랜딩을 포함하는 root 레이아웃 내에서 fallback 컴포넌트가 렌더링될 root `error.js`를 정의하는 것이 좋다.

`app/global-error.tsx`

```tsx
"use client";

export default function GlobalError({ error, reset }: { error: Error & { digest?: string }; reset: () => void }) {
  return (
    <html>
      <body>
        <h2>Something went wrong!</h2>
        <button onClick={() => reset()}>Try again</button>
      </body>
    </html>
  );
}
```

---

#### 서버 오류 처리

서버 컴포넌트 내부에서 오류가 발생하면 Next.js는 `Error` 개체(프로덕션에서 민감한 오류 정보가 제거된)를 가장 가까운 `error.js` 파일에 `error` prop으로 전달한다.

#### 민감한 오류 정보 보호

오류에 포함된 잠재적 민간 세부 정보가 클라이언트에 유출되는 것을 방지하기 위해 프로덕션 중에 클라이언트에 전달되는 `Error` 개체에는 일반 `message`와 `digest` 속성만 포함된다. 이것은 보안 예방 조치다.

`message` 속성에는 오류에 대한 일반 메시지가 포함되어 있고 `digest` 속성에는 서버 측 로그의 해당 오류와 일치시키는 데 사용할 수 있는 자동으로 생성된 오류 해시가 포함되어 있다.

개발 중에 클라이언트에 전달된 `Error` 개체는 직렬화되고 보다 쉬운 디버깅을 위해 원래 오류 메시지를 포함한다.
